File: sieve2.cpp
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
int main() {
  printf("CPP (GNU c++)\n");
  const int Expected = 15485863;
  const int N = 16000000; const int FINDPRIME = 1000000;
  const int N_sqrt = (int) sqrt((double)N);
  int count = 1;
  int i;
  unsigned char *b = (unsigned char*)calloc(N, sizeof(char));
  for (i = 3; i <= N_sqrt; i += 2) {
    if ( b[i] ) continue;
    count++;
    for (int k = i*i; k <= N; k += 2 * i) b[k]=1;
  }
  for (; i <= N; i += 2) {
    if ( ! b[i] ) {
      count++;
    }
    if (count >= FINDPRIME) break;
  }
  printf("Total Primes = %d\n", count);
  printf("Max Prime = %d\n", i);
  if (i != Expected) printf("ERROR: Largest Prime was incorrect. Should be: %d\n", Expected);
}

File: sieve2.php
<?php
print "PHP\n";
$Expected = 15485863;
$N = 16000000; $FINDPRIME = 1000000;
$count = 1; $N_sqrt = sqrt($N);
$b = array();
for ($i = 3; $i < $N; $i += 2) $b[$i] = 1;
for ($i = 3; $i <= $N_sqrt; $i += 2) {
    if ($b[$i] == 0) continue;
    $count++;
    for ( $k = $i*$i;$k <= $N ; $k += 2*$i) $b[$k] = 0;
}
for (;$i <= $N; $i += 2) {
    if ($b[$i] == 0) continue;
    $count++;
    if ($count >= $FINDPRIME) break;
}
print "Total Primes  = $count\n";
print "Max Prime = $i\n";
if ($i != $Expected) print "ERROR: Largest Prime was incorrect. Should be: $Expected\n";
?>

